## Polars

### ¿Qué es Polars? 
#### Polars es una librería diseñada para trabajar con datos tabulares (DataFrames). Tiene como principal característica la capacidad de procesar grandes volúmenes de datos de forma rápida y eficiente, gracias a que maximiza el uso de todos los cores disponibles en un ordenador.
### La librería se importa como:


In [2]:
import polars as pl

#### Polars puede trabajar con distintos formatos de datos (csv, parquet, json, DBs, etc.). 

In [3]:
# Lectura de datos (csv)
df=pl.read_csv('CIA_SEGURO_2023.csv')

print(f"Tipo de objeto: {type(df)}")
print(f"Dimensiones de los datos: {df.shape}")
display(df.head())

Tipo de objeto: <class 'polars.dataframe.frame.DataFrame'>
Dimensiones de los datos: (7292073, 55)


shape: (5, 55)
┌──────────────┬──────────────┬────────────┬──────────────┬───┬──────────────┬─────────────┬─────────────┬─────────────┐
│ PERIODO_PRES ┆ FECHA_DIGITA ┆ FOLIO_BONO ┆ ID_RUT_TITUL ┆ … ┆ CODATE_TRAZA ┆ GLS_CODATE_ ┆ CODATE_TRAZ ┆ GLS_CODATE_ │
│ TACION       ┆ CION         ┆ ---        ┆ AR           ┆   ┆ DORA_2       ┆ TRAZADORA_2 ┆ ADORA_3     ┆ TRAZADORA_3 │
│ ---          ┆ ---          ┆ i64        ┆ ---          ┆   ┆ ---          ┆ ---         ┆ ---         ┆ ---         │
│ i64          ┆ str          ┆            ┆ i64          ┆   ┆ i64          ┆ str         ┆ i64         ┆ str         │
╞══════════════╪══════════════╪════════════╪══════════════╪═══╪══════════════╪═════════════╪═════════════╪═════════════╡
│ 202304       ┆ 1900-01-01   ┆ 119557462  ┆ 93000581     ┆ … ┆ 0            ┆ null        ┆ 0           ┆ null        │
│              ┆ 00:00:00.000 ┆            ┆              ┆   ┆              ┆             ┆             ┆             │
│ 202304       ┆ 1900-01-01   ┆ 119557603  ┆ 85287436     ┆ … ┆ 0            ┆ null        ┆ 0           ┆ null        │
│              ┆ 00:00:00.000 ┆            ┆              ┆   ┆              ┆             ┆             ┆             │
│ 202304       ┆ 1900-01-01   ┆ 119557699  ┆ 71892611     ┆ … ┆ 0            ┆ null        ┆ 0           ┆ null        │
│              ┆ 00:00:00.000 ┆            ┆              ┆   ┆              ┆             ┆             ┆             │
│ 202304       ┆ 1900-01-01   ┆ 119557718  ┆ 86173251     ┆ … ┆ 0            ┆ null        ┆ 0           ┆ null        │
│              ┆ 00:00:00.000 ┆            ┆              ┆   ┆              ┆             ┆             ┆             │
│ 202304       ┆ 1900-01-01   ┆ 119557994  ┆ 39361821     ┆ … ┆ 0            ┆ null        ┆ 0           ┆ null        │
│              ┆ 00:00:00.000 ┆            ┆              ┆   ┆              ┆             ┆             ┆             │
└──────────────┴──────────────┴────────────┴──────────────┴───┴──────────────┴─────────────┴─────────────┴─────────────┘

#### La primera cosa que podemos notar es que, la función de lectura csv, es muy similar a la syntax de Pandas .read_csv() y que también nos permite pasar argumentos como sep, columns, skip_rows, etc.
#### El objeto devuelto es de tipo polars.dataframe.frame.DataFrame, contiene 7292073 filas y 55 columnas.
#### Al igual que en Pandas, los polars.DataFrame tiene atributos capaces de informarnos del shape, schema, columns y dtypes. 
#### Los polars.DataFrame también poseen atributos y métodos que permiten explorar y visualizar los datos de forma rápida. En el código anterior hemos llamado al método .head() que devuelve por defecto las 5 primeras líneas del DataFrame. Otros métodos para una rápida exploración son .tail() y .sample().
#### Es interesante destacar que, cuando hacemos un display del DataFrame, se muestra debajo del nombre de cada columna el tipo de datos.
#### A diferencia de los DataFrame de Pandas, los polars.DataFrame no tiene un index asociado a las filas.

### Atributos de un DataFrame

In [4]:
n_chars = 50

# df.columns
# ==============================================================================
print("="*n_chars)
print("df.columns: listado de columnas de un DataFrame")
print(df.columns)
print("")

# df.types
# ==============================================================================
print("="*n_chars)
print("df.types: tipos de datos de las columnas de un DataFrame")
print(df.dtypes)
print(f"Tipos de datos únicos: {set(df.dtypes)}")
print("")

# df.schema
# ==============================================================================
print("="*n_chars)
print("df.schema: Diccionario con las columnas y sus tipos de datos")
print(df.schema)
print("")

# df.shape
# ==============================================================================
print("="*n_chars)
print("df.shape : filas x columnas de un DataFrame")
print(df.shape)
print("")

# df.height
# ==============================================================================
print("="*n_chars)
print("df.height : número de filas de un DataFrame")
print(df.height)
print("")

# df.width
# ==============================================================================
print("="*n_chars)
print("df.width : número de columnas de un DataFrame")
print(df.width)
print("")

df.columns: listado de columnas de un DataFrame
['PERIODO_PRESTACION', 'FECHA_DIGITACION', 'FOLIO_BONO', 'ID_RUT_TITULAR', 'ID_RUT_BENEFICIARIO', 'CODIGO_CIE', 'GLOSA_CIE', 'PERIODO_RED', 'FOLIO_CAEC', 'COD_DESCRIPCION_CAEC', 'GLS_DESCRIPCION_CAEC', 'SEXO_BENEFICIARIO', 'FECHA_NACIMIENTO', 'CANTIDADES_CARGA', 'HOLDING_PRESTADOR', 'TIPO_COBERTURA', 'DERIVADO_URGENCIA', 'CODIGO_PLAN_ISAPRE', 'GLOSA_PLAN_ISAPRE', 'PACTADO_FINAL', 'PACTADO_PLAN', 'COD_REGION', 'GLOSA_REGION', 'FECHA_INGRESO_ISAPRE', 'PERIODO_INGRESO_PLAN', 'FECHA_VIGENCIA_PLAN', 'FECHA_OCURRENCIA_GASTO', 'FECHA_REEMB_ISAPRE', 'COBRADO', 'BONIFICADO', 'DEDUCIBLE_CAEC', 'DEDUCIBLE_GES', 'MONTO_CAEC', 'COPAGO', 'RUT_PRESTADOR', 'NOMBRE_PRESTADOR', 'PRESTACION', 'SUB_TIPO_PRESTACION', 'GRUPO', 'SUBGRUPO', 'TIPO_BONO', 'GLS_TIPO_BONO', 'SOLICITUD_TRATAMIENTO', 'RUBRO', 'NOMBRE_RUBRO', 'ACTOR', 'NOMBRE_ACTOR', 'CODATE_DETALLE', 'GLS_CODATE_DETALLE', 'CODATE_TRAZADORA', 'GLS_CODATE_TRAZADORA', 'CODATE_TRAZADORA_2', 'GLS_CODATE_TR

### Métricas y valores descriptivos

##### Los polars.DataFrame tienen distintos métodos capaces de generar métricas descriptivas de los datos. Entre ellos destaca el método .null_count() devuelve un DataFrame con el total de valores nulos por columna.

In [5]:
# df.null_count()

print("df.null_count(): Devuelve un DataFrame con el total de valores nulos por columna")
display(df.null_count())

df.null_count(): Devuelve un DataFrame con el total de valores nulos por columna


shape: (1, 55)
┌──────────────┬──────────────┬────────────┬──────────────┬───┬──────────────┬─────────────┬─────────────┬─────────────┐
│ PERIODO_PRES ┆ FECHA_DIGITA ┆ FOLIO_BONO ┆ ID_RUT_TITUL ┆ … ┆ CODATE_TRAZA ┆ GLS_CODATE_ ┆ CODATE_TRAZ ┆ GLS_CODATE_ │
│ TACION       ┆ CION         ┆ ---        ┆ AR           ┆   ┆ DORA_2       ┆ TRAZADORA_2 ┆ ADORA_3     ┆ TRAZADORA_3 │
│ ---          ┆ ---          ┆ u32        ┆ ---          ┆   ┆ ---          ┆ ---         ┆ ---         ┆ ---         │
│ u32          ┆ u32          ┆            ┆ u32          ┆   ┆ u32          ┆ u32         ┆ u32         ┆ u32         │
╞══════════════╪══════════════╪════════════╪══════════════╪═══╪══════════════╪═════════════╪═════════════╪═════════════╡
│ 0            ┆ 1052200      ┆ 0          ┆ 0            ┆ … ┆ 1301552      ┆ 7251737     ┆ 1301552     ┆ 7284648     │
└──────────────┴──────────────┴────────────┴──────────────┴───┴──────────────┴─────────────┴─────────────┴─────────────┘

### Polar Expressions
#### Las Expressions son funciones/métodos utilizados a la hora de realizar operaciones con datos en Polars (e.g., selección, creación y manipulación de columnas, aplicación de filtros, entre otros). Tienen como entrada una serie y como salida otra serie, y són, por definición, un mapeo entre séries, lo que nos permite encadenarlas. Además, Polars es capaz de automatizar la ejecución de las expressions en paralelo siempre que sea posible (cuando se trabaja con múltiples columnas, por ejemplo) lo que hace de las Expressions algo muy potente.

#### Selección de columnas: vamos ahora ver un ejemplo de cómo seleccionar columna(s) utilizando a Polars Expressions:

In [6]:
# .select(): Método utilizado para seleccionar una o más columnas
# ==============================================================================
df.select(pl.col('ID_RUT_TITULAR'))

shape: (7_292_073, 1)
┌────────────────┐
│ ID_RUT_TITULAR │
│ ---            │
│ i64            │
╞════════════════╡
│ 93000581       │
│ 85287436       │
│ 71892611       │
│ 86173251       │
│ 39361821       │
│ …              │
│ 93824811       │
│ 72319961       │
│ 28228646       │
│ 33871981       │
│ 48893271       │
└────────────────┘

In [7]:
type(df.select(pl.col('ID_RUT_TITULAR')))

polars.dataframe.frame.DataFrame

##### Cómo podemos observar en este ejemplo, hemos utilizado el método .select() para seleccionar una columna específica de nuestro DataFrame ('ID_RUT_TITULAR'). Por defecto, este método nos retorna otro DataFrame (polars.dataframe.frame.DataFrame)
#### Para seleccionar columnas podemos pasar una lista con las columnas deseadas:

In [8]:
df.select([pl.col('ID_RUT_TITULAR'),pl.col('SEXO_BENEFICIARIO')])

shape: (7_292_073, 2)
┌────────────────┬───────────────────┐
│ ID_RUT_TITULAR ┆ SEXO_BENEFICIARIO │
│ ---            ┆ ---               │
│ i64            ┆ str               │
╞════════════════╪═══════════════════╡
│ 93000581       ┆ F                 │
│ 85287436       ┆ M                 │
│ 71892611       ┆ F                 │
│ 86173251       ┆ M                 │
│ 39361821       ┆ F                 │
│ …              ┆ …                 │
│ 93824811       ┆ F                 │
│ 72319961       ┆ F                 │
│ 28228646       ┆ M                 │
│ 33871981       ┆ M                 │
│ 48893271       ┆ F                 │
└────────────────┴───────────────────┘

#### Al utilizar la expresión pl.col() dentro del contexto .select(), uno puede realizar más manipulaciones en la columna. De hecho, puedes usar tantas expresiones en la columna como quieras, lo que te permite realizar varias operaciones. 

#### Creando una secuencia de tareas: en esta sección vamos a entender de manera muy sencilla cómo podemos crear una secuencia de pasos para manipular nuestros datos. Estos serán los pasos que vamos a implementar:
###### Seleccionar y renombrar columnas
###### Crear una nueva columna
###### Aplicar un filtro a una determinada columna
###### Realizar una ordenación

In [9]:
# 1. Seleccionar y renombrar columnas
# ==============================================================================

nuevo_df = (df.select([
            pl.col('SEXO_BENEFICIARIO').alias('sexo_benef'),
            pl.col('CANTIDADES_CARGA').alias('nro_cargas'),
            pl.col('FECHA_NACIMIENTO').alias('fecha_cumple'),
            pl.col('COBRADO').alias('cobrado'),
            pl.col('BONIFICADO').alias('bonificado'),
            pl.col('GRUPO').alias('grupo')]))
nuevo_df.head()

shape: (5, 6)
┌────────────┬────────────┬─────────────────────────┬─────────┬────────────┬──────────────┐
│ sexo_benef ┆ nro_cargas ┆ fecha_cumple            ┆ cobrado ┆ bonificado ┆ grupo        │
│ ---        ┆ ---        ┆ ---                     ┆ ---     ┆ ---        ┆ ---          │
│ str        ┆ i64        ┆ str                     ┆ i64     ┆ i64        ┆ str          │
╞════════════╪════════════╪═════════════════════════╪═════════╪════════════╪══════════════╡
│ F          ┆ 0          ┆ 1995-09-02 00:00:00.000 ┆ 32660   ┆ 26128      ┆ IMAGENOLOGIA │
│ M          ┆ 0          ┆ 1991-02-07 00:00:00.000 ┆ 8160    ┆ 8160       ┆ EXAMENES     │
│ F          ┆ 0          ┆ 1982-05-28 00:00:00.000 ┆ 820     ┆ 820        ┆ EXAMENES     │
│ M          ┆ 0          ┆ 1991-06-04 00:00:00.000 ┆ 40154   ┆ 28108      ┆ CONSULTAS    │
│ F          ┆ 0          ┆ 1975-06-20 00:00:00.000 ┆ 1200    ┆ 1200       ┆ OTRAS        │
└────────────┴────────────┴─────────────────────────┴─────────┴────────────┴──────────────┘

##### Cómo podemos ver, una manera de renombrar las columnas es a través de la expresión alias() que puede ser encadenada a cada una de las columnas dentro del select().

In [10]:
# 2. Creación de nuevas columnas: unit_cost, execution_date
# ==============================================================================
from datetime import datetime
nuevo_df = (nuevo_df.with_columns([
        (pl.col('cobrado')+pl.col('bonificado')).alias('TOTAL'),
        (pl.lit(datetime.now().date()).alias('fecha_ejecucion')),]))
nuevo_df.head()

shape: (5, 8)
┌────────────┬────────────┬─────────────────────────┬─────────┬────────────┬──────────────┬───────┬─────────────────┐
│ sexo_benef ┆ nro_cargas ┆ fecha_cumple            ┆ cobrado ┆ bonificado ┆ grupo        ┆ TOTAL ┆ fecha_ejecucion │
│ ---        ┆ ---        ┆ ---                     ┆ ---     ┆ ---        ┆ ---          ┆ ---   ┆ ---             │
│ str        ┆ i64        ┆ str                     ┆ i64     ┆ i64        ┆ str          ┆ i64   ┆ date            │
╞════════════╪════════════╪═════════════════════════╪═════════╪════════════╪══════════════╪═══════╪═════════════════╡
│ F          ┆ 0          ┆ 1995-09-02 00:00:00.000 ┆ 32660   ┆ 26128      ┆ IMAGENOLOGIA ┆ 58788 ┆ 2024-02-14      │
│ M          ┆ 0          ┆ 1991-02-07 00:00:00.000 ┆ 8160    ┆ 8160       ┆ EXAMENES     ┆ 16320 ┆ 2024-02-14      │
│ F          ┆ 0          ┆ 1982-05-28 00:00:00.000 ┆ 820     ┆ 820        ┆ EXAMENES     ┆ 1640  ┆ 2024-02-14      │
│ M          ┆ 0          ┆ 1991-06-04 00:00:00.000 ┆ 40154   ┆ 28108      ┆ CONSULTAS    ┆ 68262 ┆ 2024-02-14      │
│ F          ┆ 0          ┆ 1975-06-20 00:00:00.000 ┆ 1200    ┆ 1200       ┆ OTRAS        ┆ 2400  ┆ 2024-02-14      │
└────────────┴────────────┴─────────────────────────┴─────────┴────────────┴──────────────┴───────┴─────────────────┘

##### Para crear columnas utilizamos la expresión with_columns() donde podemos crear varias columnas a la vez. pl.lit() se utiliza cuando queremos propagar una constante para todo el DataFrame. En este ejemplo hemos creado una fecha de ejecución que ha sido propagada en todo el DataFrame en la nueva columna fecha_ejecucion que fue automáticamente inferida cómo tipo date por Polars.

In [11]:
# 3. Filtrar filas: GRUPO = 'EXAMENES' or  'CONSULTAS'
# ==============================================================================
nuevo_df = (nuevo_df.filter((pl.col('grupo') == 'EXAMENES') | (pl.col('grupo') == 'CONSULTAS')))
nuevo_df.head()

shape: (5, 8)
┌────────────┬────────────┬─────────────────────────┬─────────┬────────────┬───────────┬───────┬─────────────────┐
│ sexo_benef ┆ nro_cargas ┆ fecha_cumple            ┆ cobrado ┆ bonificado ┆ grupo     ┆ TOTAL ┆ fecha_ejecucion │
│ ---        ┆ ---        ┆ ---                     ┆ ---     ┆ ---        ┆ ---       ┆ ---   ┆ ---             │
│ str        ┆ i64        ┆ str                     ┆ i64     ┆ i64        ┆ str       ┆ i64   ┆ date            │
╞════════════╪════════════╪═════════════════════════╪═════════╪════════════╪═══════════╪═══════╪═════════════════╡
│ M          ┆ 0          ┆ 1991-02-07 00:00:00.000 ┆ 8160    ┆ 8160       ┆ EXAMENES  ┆ 16320 ┆ 2024-02-14      │
│ F          ┆ 0          ┆ 1982-05-28 00:00:00.000 ┆ 820     ┆ 820        ┆ EXAMENES  ┆ 1640  ┆ 2024-02-14      │
│ M          ┆ 0          ┆ 1991-06-04 00:00:00.000 ┆ 40154   ┆ 28108      ┆ CONSULTAS ┆ 68262 ┆ 2024-02-14      │
│ M          ┆ 1          ┆ 1984-10-14 00:00:00.000 ┆ 1310    ┆ 771        ┆ EXAMENES  ┆ 2081  ┆ 2024-02-14      │
│ F          ┆ 1          ┆ 1961-12-31 00:00:00.000 ┆ 7921    ┆ 5545       ┆ EXAMENES  ┆ 13466 ┆ 2024-02-14      │
└────────────┴────────────┴─────────────────────────┴─────────┴────────────┴───────────┴───────┴─────────────────┘

In [12]:
df.head()

shape: (5, 55)
┌──────────────┬──────────────┬────────────┬──────────────┬───┬──────────────┬─────────────┬─────────────┬─────────────┐
│ PERIODO_PRES ┆ FECHA_DIGITA ┆ FOLIO_BONO ┆ ID_RUT_TITUL ┆ … ┆ CODATE_TRAZA ┆ GLS_CODATE_ ┆ CODATE_TRAZ ┆ GLS_CODATE_ │
│ TACION       ┆ CION         ┆ ---        ┆ AR           ┆   ┆ DORA_2       ┆ TRAZADORA_2 ┆ ADORA_3     ┆ TRAZADORA_3 │
│ ---          ┆ ---          ┆ i64        ┆ ---          ┆   ┆ ---          ┆ ---         ┆ ---         ┆ ---         │
│ i64          ┆ str          ┆            ┆ i64          ┆   ┆ i64          ┆ str         ┆ i64         ┆ str         │
╞══════════════╪══════════════╪════════════╪══════════════╪═══╪══════════════╪═════════════╪═════════════╪═════════════╡
│ 202304       ┆ 1900-01-01   ┆ 119557462  ┆ 93000581     ┆ … ┆ 0            ┆ null        ┆ 0           ┆ null        │
│              ┆ 00:00:00.000 ┆            ┆              ┆   ┆              ┆             ┆             ┆             │
│ 202304       ┆ 1900-01-01   ┆ 119557603  ┆ 85287436     ┆ … ┆ 0            ┆ null        ┆ 0           ┆ null        │
│              ┆ 00:00:00.000 ┆            ┆              ┆   ┆              ┆             ┆             ┆             │
│ 202304       ┆ 1900-01-01   ┆ 119557699  ┆ 71892611     ┆ … ┆ 0            ┆ null        ┆ 0           ┆ null        │
│              ┆ 00:00:00.000 ┆            ┆              ┆   ┆              ┆             ┆             ┆             │
│ 202304       ┆ 1900-01-01   ┆ 119557718  ┆ 86173251     ┆ … ┆ 0            ┆ null        ┆ 0           ┆ null        │
│              ┆ 00:00:00.000 ┆            ┆              ┆   ┆              ┆             ┆             ┆             │
│ 202304       ┆ 1900-01-01   ┆ 119557994  ┆ 39361821     ┆ … ┆ 0            ┆ null        ┆ 0           ┆ null        │
│              ┆ 00:00:00.000 ┆            ┆              ┆   ┆              ┆             ┆             ┆             │
└──────────────┴──────────────┴────────────┴──────────────┴───┴──────────────┴─────────────┴─────────────┴─────────────┘

#### La principal ventaja de trabajar con las Polars Expressions es poder concatenarlas dentro de un solo bloque de código. Abajo, vamos a crear una función de preprocesado para ejemplificar esta secuencia de tareas:


In [13]:
# Función de preprocesado
# ==============================================================================
def agrupado_funciones(raw_data):
    df_funciones_agrupadas=raw_data.select([
                pl.col('SEXO_BENEFICIARIO').alias('sexo_benef'),
                pl.col('CANTIDADES_CARGA').alias('nro_cargas'),
                pl.col('FECHA_NACIMIENTO').alias('fecha_cumple'), pl.col('COBRADO').alias('cobrado'),
                pl.col('BONIFICADO').alias('bonificado'),
                pl.col('GRUPO').alias('grupo')]).with_columns([
                (pl.col('cobrado')+pl.col('bonificado')).alias('TOTAL'),
                (pl.lit(datetime.now().date()).alias('fecha_ejecucion')),]).filter((pl.col('grupo') == 'EXAMENES') | (pl.col('grupo') == 'CONSULTAS'))

    return df_funciones_agrupadas
df_procesado=agrupado_funciones(raw_data=df)
display(df_procesado.head())

shape: (5, 8)
┌────────────┬────────────┬─────────────────────────┬─────────┬────────────┬───────────┬───────┬─────────────────┐
│ sexo_benef ┆ nro_cargas ┆ fecha_cumple            ┆ cobrado ┆ bonificado ┆ grupo     ┆ TOTAL ┆ fecha_ejecucion │
│ ---        ┆ ---        ┆ ---                     ┆ ---     ┆ ---        ┆ ---       ┆ ---   ┆ ---             │
│ str        ┆ i64        ┆ str                     ┆ i64     ┆ i64        ┆ str       ┆ i64   ┆ date            │
╞════════════╪════════════╪═════════════════════════╪═════════╪════════════╪═══════════╪═══════╪═════════════════╡
│ M          ┆ 0          ┆ 1991-02-07 00:00:00.000 ┆ 8160    ┆ 8160       ┆ EXAMENES  ┆ 16320 ┆ 2024-02-14      │
│ F          ┆ 0          ┆ 1982-05-28 00:00:00.000 ┆ 820     ┆ 820        ┆ EXAMENES  ┆ 1640  ┆ 2024-02-14      │
│ M          ┆ 0          ┆ 1991-06-04 00:00:00.000 ┆ 40154   ┆ 28108      ┆ CONSULTAS ┆ 68262 ┆ 2024-02-14      │
│ M          ┆ 1          ┆ 1984-10-14 00:00:00.000 ┆ 1310    ┆ 771        ┆ EXAMENES  ┆ 2081  ┆ 2024-02-14      │
│ F          ┆ 1          ┆ 1961-12-31 00:00:00.000 ┆ 7921    ┆ 5545       ┆ EXAMENES  ┆ 13466 ┆ 2024-02-14      │
└────────────┴────────────┴─────────────────────────┴─────────┴────────────┴───────────┴───────┴─────────────────┘

### Lazy evaluation
Hasta el momento, todo lo que hemos ejecutado en nuestro DataFrame (df:polars.DataFrame) se ejecutó en modo Eager, es decir, de manera instantánea de acuerdo con lo que hemos definido en las celdas, tal cómo ocurriría si fuera un DataFrame de Pandas.

Sin embargo, Polars también tiene una forma Lazy para evaluar/ejecutar las instrucciones de código. Esta forma Lazy permite a Polars evaluar la sintaxis de nuestro código/expresión, optimizarla y finalmente ejecutarla dentro de su engine. Esto permite, en general, mejorar aún más el rendimiento y optimizar el uso de la memoria. En modo Lazy, Polars crea y realiza un seguimiento de nuestro código en un plan lógico donde es capaz de optimizar y reordenar cada tarea antes de ejecutarlo. Vamos ahora a explorar un poco el universo de las Lazy Evaluations.

In [14]:
lazy_df = df.lazy()
print(f"Object type: {type(lazy_df)}")

Object type: <class 'polars.lazyframe.frame.LazyFrame'>


In [15]:
print(lazy_df.shape)

AttributeError: 'LazyFrame' object has no attribute 'shape'

In [16]:
print(lazy_df)

naive plan: (run LazyFrame.explain(optimized=True) to see the optimized plan)

DF ["PERIODO_PRESTACION", "FECHA_DIGITACION", "FOLIO_BONO", "ID_RUT_TITULAR"]; PROJECT */55 COLUMNS; SELECTION: "None"


En primer lugar hemos creado un polars.LazyFrame a partir de nuestro polars.DataFrame inicial a través del método .lazy().

Un LazyFrame es un objeto sobre el cual Polars es capaz de crear un plan de ejecución y optimizar queries, volviendo así más eficiente el proceso de transformación y manipulación de los datos.

Existen también maneras de realizar una lectura de datos en modo Lazy directamente, como por ejemplo: pl.scan_csv, pl.scan_parquet, etc.

Al intentar mostrar el shape de nuestro lazy dataframe se obtiene un error porque no todos los métodos disponibles para los polars.DataFrame están disponibles también para polars.LazyFrame, entre ellos el .shape()

Al llamar a nuestro polars.LazyFrame directamente, lo que nos devuelve es el plan de ejecución de datos. En este caso, no está haciendo nada en especial ya que solamente estamos cargando los datos. Pero es importante entender que, a medida que vamos trabajando con este DataFrame, el plan de ejecución también va cambiando y cuando necesitemos ejecutarlo, Polars es capaz de optimizarlo y paralelizar la ejecución de las tareas siempre que sea posible.

Vamos ahora llamar a la misma función de preprocesado pero utilizando nuestro polars.LazyFrame:

In [17]:
lazy_procesado_df = agrupado_funciones(raw_data=lazy_df)
lazy_procesado_df

<LazyFrame [8 cols, {"sexo_benef": String … "fecha_ejecucion": Date}] at 0x21D32D69510>

In [18]:
#Muestra el lazy_procesado_df optimizado

lazy_procesado_df.explain(optimized=True) 

' WITH_COLUMNS:\n [[(col("cobrado")) + (col("bonificado"))].alias("TOTAL"), 2024-02-14.alias("fecha_ejecucion")]\n   SELECT [col("SEXO_BENEFICIARIO").alias("sexo_benef"), col("CANTIDADES_CARGA").alias("nro_cargas"), col("FECHA_NACIMIENTO").alias("fecha_cumple"), col("COBRADO").alias("cobrado"), col("BONIFICADO").alias("bonificado"), col("GRUPO").alias("grupo")] FROM\n    DF ["PERIODO_PRESTACION", "FECHA_DIGITACION", "FOLIO_BONO", "ID_RUT_TITULAR"]; PROJECT 6/55 COLUMNS; SELECTION: "[([(col(\\"GRUPO\\")) == (String(EXAMENES))]) | ([(col(\\"GRUPO\\")) == (String(CONSULTAS))])]"'

##### Nuestra función se ha ejecutado sin problemas y nos ha devuelto otro objeto LazyFrame. Al intentar visualizar este objeto, cómo ya sabemos, nos salta su plan de ejecución no optimizado (NAIVE QUERY PLAN). Podemos visualizar los planes de ejecución (NAIVE QUERY PLAN/OPTIMIZED QUERY PLAN) de nuestro DataFrame a través del método: lazy_procesado_df.show_graph()

In [21]:
#lazy_procesado_df.show_graph()

###### Si en algún momento de nuestro +pipeline+ necesitamos realmente visualizar el pl.LazyDataFrame con sus datos, podemos utilizar métodos como collect() y fetch() que nos van a retornar un pl.DataFrame. El método fetch() es muy indicado para hacer debugging ya que es una operación rápida y que limita el número final de filas devueltas.

In [22]:
# .fetch(): devuelve un número limitado de líneas para hacer debugging
# ==============================================================================
print(type(lazy_procesado_df.fetch(20)))
lazy_procesado_df.fetch(20)

<class 'polars.dataframe.frame.DataFrame'>


shape: (20, 8)
┌────────────┬────────────┬─────────────────────────┬─────────┬────────────┬───────────┬────────┬─────────────────┐
│ sexo_benef ┆ nro_cargas ┆ fecha_cumple            ┆ cobrado ┆ bonificado ┆ grupo     ┆ TOTAL  ┆ fecha_ejecucion │
│ ---        ┆ ---        ┆ ---                     ┆ ---     ┆ ---        ┆ ---       ┆ ---    ┆ ---             │
│ str        ┆ i64        ┆ str                     ┆ i64     ┆ i64        ┆ str       ┆ i64    ┆ date            │
╞════════════╪════════════╪═════════════════════════╪═════════╪════════════╪═══════════╪════════╪═════════════════╡
│ M          ┆ 0          ┆ 1991-02-07 00:00:00.000 ┆ 8160    ┆ 8160       ┆ EXAMENES  ┆ 16320  ┆ 2024-02-14      │
│ F          ┆ 0          ┆ 1982-05-28 00:00:00.000 ┆ 820     ┆ 820        ┆ EXAMENES  ┆ 1640   ┆ 2024-02-14      │
│ M          ┆ 0          ┆ 1991-06-04 00:00:00.000 ┆ 40154   ┆ 28108      ┆ CONSULTAS ┆ 68262  ┆ 2024-02-14      │
│ M          ┆ 1          ┆ 1984-10-14 00:00:00.000 ┆ 1310    ┆ 771        ┆ EXAMENES  ┆ 2081   ┆ 2024-02-14      │
│ F          ┆ 1          ┆ 1961-12-31 00:00:00.000 ┆ 7921    ┆ 5545       ┆ EXAMENES  ┆ 13466  ┆ 2024-02-14      │
│ …          ┆ …          ┆ …                       ┆ …       ┆ …          ┆ …         ┆ …      ┆ …               │
│ F          ┆ 0          ┆ 1988-01-21 00:00:00.000 ┆ 64484   ┆ 58036      ┆ CONSULTAS ┆ 122520 ┆ 2024-02-14      │
│ F          ┆ 2          ┆ 1963-03-15 00:00:00.000 ┆ 6770    ┆ 4672       ┆ EXAMENES  ┆ 11442  ┆ 2024-02-14      │
│ M          ┆ 1          ┆ 2004-05-03 00:00:00.000 ┆ 23691   ┆ 7875       ┆ EXAMENES  ┆ 31566  ┆ 2024-02-14      │
│ M          ┆ 1          ┆ 1966-01-17 00:00:00.000 ┆ 7650    ┆ 6120       ┆ EXAMENES  ┆ 13770  ┆ 2024-02-14      │
│ M          ┆ 2          ┆ 2003-11-19 00:00:00.000 ┆ 71464   ┆ 50025      ┆ CONSULTAS ┆ 121489 ┆ 2024-02-14      │
└────────────┴────────────┴─────────────────────────┴─────────┴────────────┴───────────┴────────┴─────────────────┘

In [23]:
# .collect(): devuelve todo el DataFrame
# ==============================================================================
print(type(lazy_procesado_df.collect()))
lazy_procesado_df.collect()

<class 'polars.dataframe.frame.DataFrame'>


shape: (4_488_142, 8)
┌────────────┬────────────┬─────────────────────────┬─────────┬────────────┬───────────┬───────┬─────────────────┐
│ sexo_benef ┆ nro_cargas ┆ fecha_cumple            ┆ cobrado ┆ bonificado ┆ grupo     ┆ TOTAL ┆ fecha_ejecucion │
│ ---        ┆ ---        ┆ ---                     ┆ ---     ┆ ---        ┆ ---       ┆ ---   ┆ ---             │
│ str        ┆ i64        ┆ str                     ┆ i64     ┆ i64        ┆ str       ┆ i64   ┆ date            │
╞════════════╪════════════╪═════════════════════════╪═════════╪════════════╪═══════════╪═══════╪═════════════════╡
│ M          ┆ 0          ┆ 1991-02-07 00:00:00.000 ┆ 8160    ┆ 8160       ┆ EXAMENES  ┆ 16320 ┆ 2024-02-14      │
│ F          ┆ 0          ┆ 1982-05-28 00:00:00.000 ┆ 820     ┆ 820        ┆ EXAMENES  ┆ 1640  ┆ 2024-02-14      │
│ M          ┆ 0          ┆ 1991-06-04 00:00:00.000 ┆ 40154   ┆ 28108      ┆ CONSULTAS ┆ 68262 ┆ 2024-02-14      │
│ M          ┆ 1          ┆ 1984-10-14 00:00:00.000 ┆ 1310    ┆ 771        ┆ EXAMENES  ┆ 2081  ┆ 2024-02-14      │
│ F          ┆ 1          ┆ 1961-12-31 00:00:00.000 ┆ 7921    ┆ 5545       ┆ EXAMENES  ┆ 13466 ┆ 2024-02-14      │
│ …          ┆ …          ┆ …                       ┆ …       ┆ …          ┆ …         ┆ …     ┆ …               │
│ M          ┆ 3          ┆ 2017-06-23 00:00:00.000 ┆ 40000   ┆ 28000      ┆ CONSULTAS ┆ 68000 ┆ 2024-02-14      │
│ F          ┆ 1          ┆ 1948-08-30 00:00:00.000 ┆ 47945   ┆ 21345      ┆ CONSULTAS ┆ 69290 ┆ 2024-02-14      │
│ F          ┆ 1          ┆ 1989-10-25 00:00:00.000 ┆ 48930   ┆ 42691      ┆ CONSULTAS ┆ 91621 ┆ 2024-02-14      │
│ F          ┆ 0          ┆ 1996-05-12 00:00:00.000 ┆ 27000   ┆ 17788      ┆ CONSULTAS ┆ 44788 ┆ 2024-02-14      │
│ M          ┆ 1          ┆ 1956-09-21 00:00:00.000 ┆ 60000   ┆ 27698      ┆ CONSULTAS ┆ 87698 ┆ 2024-02-14      │
└────────────┴────────────┴─────────────────────────┴─────────┴────────────┴───────────┴───────┴─────────────────┘

### Agrupación de Datos - GroupBy:
Vamos a agrupar nuestros datos preprocesados a nivel del "region_name" y calcular la cantidad total de prescripciones médicas y la media de unidades solicitadas (total_sales y avg_qt).

In [28]:
# Group by Operation
# ==============================================================================
agg_df = (
    lazy_procesado_df
    .group_by(['grupo'])
    .agg(
        [
            pl.col('cobrado').sum().alias('total_cobrado'),
            pl.col('TOTAL').mean().alias('avg_total'),
        ]
    )
    .sort('total_cobrado')
    
).collect()

agg_df

shape: (2, 3)
┌───────────┬───────────────┬──────────────┐
│ grupo     ┆ total_cobrado ┆ avg_total    │
│ ---       ┆ ---           ┆ ---          │
│ str       ┆ i64           ┆ f64          │
╞═══════════╪═══════════════╪══════════════╡
│ EXAMENES  ┆ 32464491802   ┆ 17236.481846 │
│ CONSULTAS ┆ 61570858894   ┆ 66496.362902 │
└───────────┴───────────────┴──────────────┘

##### En este ejemplo hemos utilizado las funciones .sum() y .mean() para agrupar los datos.

In [2]:
from datetime import datetime
import time
import polars as pl
start_time = time.time()
df=pl.read_csv('CIA_SEGURO_2023.csv')
print("--- %s seconds ---" % (time.time() - start_time))

--- 33.7153046131134 seconds ---


In [3]:
from datetime import datetime
import time
import pandas as pd
start_time = time.time()
df=pd.read_csv('CIA_SEGURO_2023.csv')
print("--- %s seconds ---" % (time.time() - start_time))

--- 105.29424738883972 seconds ---
